### PR TITLE
Streamline noise cancel chain

### PR DESCRIPTION
**Pull request by fedetft**
This PR does the following:

* It disables the Speex noise canceller when RNNoise is enabled as per https://github.com/mumble-voip/mumble/issues/4181
* It moves RNNoise after the echo canceller so that the input DSP chain is always Input -> Resampler -> Echo cancel -> Noise cancel -> Additional filters -> VAD -> Encoder as per https://github.com/mumble-voip/mumble/issues/4181
* It adds provision for disabling the noise canceller altogether, going in the direction of https://github.com/mumble-voip/mumble/issues/3323

Since I did not use RNNoise before, can someone who really likes RNNoise test this branch and see if it works just as well for his/her use case?

-----

The Windows installer containing the changes of this PR: 
[WindowsInstaller.zip](https://github.com/mumble-voip/mumble/files/4808046/WindowsInstaller.zip) (20.06.2020)

### CODE REVIEW

File: src/mumble/Settings.h
Comment (Krzmbrzl) [inline src/mumble/Settings.h:187]: I just wanted to request this as a change. That'll also require a little change in the UI code though. If you want, I could help on that end :)

**Comment by streaps** [comment]
Why the change? De-noise is a better term than noise cancel

**Comment by fedetft** [comment]
I'd prefer to stay away from UI code as I'm not good at it, and rather focus on the audio part. The PR is set up so that a second PR can continue from this and improve the UI. I have also left commented the line in `AudioInput::selectNoiseCancel()` that needs to be changed when that TODO is done.

Ideally, the RNNoise checkbox can be replaced by a drop-down list selecting either "Disabled, SpeexDsp, RNNoise", and also gray out the "Noise suppression slider" when SpeexDsp is not selected.

**Comment by streaps** [comment]
I also wanted to suggest a 3-choice option. Instead of drop-down maybe radio button selection. Or like this:

Screenshot_2020-05-28 RNNoise Learning Noise Suppression

**Comment by fedetft** [comment]
I suggested a drop down just for consistency with the echo cancel, but my point is that RNNoise checkbox should be replaced with something more informative.

**Comment by streaps** [comment]
It is also usually called noise suppression. I would prefer that term as "noise cancelling" is often used for active noise cancelling for headphones, which is something different.

**Comment by streaps** [comment]
A drop down is 2 clicks, radio button 1 click and it shows all options. 

Echo cancellation has something to do with the interface, but it's also audio processing. I would move that down to Audio Processing and move it on top of the list, radio selection style.

**Comment by streaps** [comment]
I would also put Audio Processing right below Interface.

Is the Noise Suppression connected to RNNoise? or is this SpeexDSP only?

**Comment by fedetft** [comment]
Considering that the DSP chain is executed in this order
`Input -> Resampler -> Echo cancel -> Noise cancel -> Additional filters -> VAD -> Encoder`
we may put the control in the UI in the same order (skipping the steps such as the resampler) that have no tunables.

Both SpeexDsp and RNNoise have a noise suppressor, this PR allows to select between them.

Based on what has been said in other topics, SpeexDsp is a good baseline noise suppressor with low CPU use, while RNNoise uses neural networks and is more CPU heavy, but also works against typing/chewing noises, which benefits some users.

**Comment by streaps** [comment]
Exactly!

The important word was missing in my question. It should have been:

Is the Noise Suppression **slider** connected to RNNoise? or is this SpeexDSP only?

**Comment by Krzmbrzl** [comment]
Okay in that case I'll have a look at the UI stuff once you're done with this PR (and someone has tested that RNNoise still works as expected) :)
I think I'd the  go ahead and push to your branch in order to have it all in a single PR (if that's okay for you)

**Comment by fedetft** [comment]
@streaps the noise suppression slider is only linked to SpeexDsp. RNNoise from what I've seen in the code has no tunables. That's why I suggested the slider be grayed out in the UI when selecting RNNoise, to prevent the user trying to tweak something that has no effect.

@Krzmbrzl I agree, let's wait for someone who uses RNNoise regularly to try this PR.
Do I need to do something to let you push to my branch?

PS: even though the option to disable the noise suppression entirely is not linked to the UI at the moment I have tested it as well and it works as expected.

**Comment by Krzmbrzl** [comment]
> Do I need to do something to let you push to my branch?

No. There is a checkbox at the right of the PR that should say "allow edits from maintainers" which is checked by default. As long as it's checked, any maintainer (including me) can push to the branch the PR has been created from (for as long as the PR is open) :)

**Comment by fedetft** [comment]
Thanks for the clarification. Checkbox is checked so it's ok.

**Comment by Chris2000SP** [comment]
I have compiled this branch and find that neither RNNoice nor Echo Cancel works properly.

Question:
Can I use the dump property to create an audio sample to dump what I have?

When I look at the audio statistics in Mumble it looks pretty broken.

**Comment by toby63** [comment]
:thinking: I would like to test this as well, but once again the problem is that most my contacts use windows or macos, so I would like to have a download for that, but Azure tells me that these builds are already not available anymore.
> Build not found.

@Krzmbrzl Could you trigger another build? And maybe you could save that build somewhere, so it doesn't get lost again?

**Comment by fedetft** [comment]
Yes @Chris2000SP, I'd like to have a look at some sample of the dumped streams.
It looks like we've found another platform-dependent issue, as I'm using this branch daily ad it works fine.
Maybe this time we'll have more luck at finding the issue. Which OS are you running mumble?

**Comment by Krzmbrzl** [comment]
> Could you trigger another build? And maybe you could save that build somewhere, so it doesn't get lost again?

Unfortunately not. If the build has been deleted already, then I can't re-run it. It could be triggered by adding another commit (or force-pushing to this branch) by @fedetft though :point_up:

**Comment by toby63** [comment]
> It could be triggered by adding another commit (or force-pushing to this branch) by @fedetft though point_up

@fedetft Thank you :+1:
@Krzmbrzl Could you save that somehow? maybe download, put somewhere here and link to it? I don't know how many people will test it, but if we want more test data, this will be the easiest way as people can simply download it.

**Comment by Krzmbrzl** [comment]
I added the installer to the first post.

**Comment by Krzmbrzl** [comment]
Any new testing results for this change? Or maybe some insights into why it seemed to not have worked for Chris2000SP?

**Comment by fedetft** [comment]
I'm keeping the branch aligned to master form time to time, and keep using it. Waiting for some more info on why for some users it doesn't work.

**Comment by toby63** [comment]
> I'm keeping the branch aligned to master form time to time, and keep using it. Waiting for some more info on why for some users it doesn't work.

I will probably try it some time next week.

Regarding the problems:
Do you have estimations for what causes the problems?
Are dependencies a potencial problem?
Or channel "management" (in pulseaudio etc.)?

@TredwellGit @felix91gr would you be interested in testing this? You seem to be regular users of RNNoise etc.

You can also download ready-for-use builds from Azure (for windows you can look at the first post, there is a copy).

**Comment by felix91gr** [comment]
I'm down, @toby63.
Tell me where the binaries are (or if I can build this, do give me instructions... I'm assuming it can't be built with cmake yet since that PR hasn't been merged right?), and I'll happily try it out. I happen to have a friend in my server who's really keen on audio noise, so he should be able to tell if my input is better with these changes :3

**Comment by toby63** [comment]
@felix91gr 
You can click on Details in the builds/checks you see on the bottom of this site.

Sadly the newer **Windows build** did not seem to succeed, but you can find a slightly older version in here(@Krzmbrzl uploaded it): https://github.com/mumble-voip/mumble/files/4808046/WindowsInstaller.zip

Here is the direct link for **Linux** (Appimage (standalone) and **Mac**:
https://dev.azure.com/Mumble-VoIP/Mumble/_build/results?buildId=2144&view=artifacts&type=publishedArtifacts
Tip: On the right side of each build there are three dots, click on them to download the builds.

For `Building`, you can use @fedetft's repo:
https://github.com/fedetft/mumble
`git clone -b rnnoise https://github.com/fedetft/mumble`

I only compiled for Linux, so I can't say how good builds for windows and mac work.

More instructions can be found here:
https://wiki.mumble.info/wiki/BuildingLinux
https://github.com/mumble-voip/mumble/blob/master/scripts/travis-ci/before_install.bash (for newest dependencies (lists Debian/Ubuntu packages))
https://github.com/mumble-voip/mumble/blob/master/scripts/travis-ci/script.bash (for build flags)

If you need a detailed guide, I can post one.

**Comment by fedetft** [comment]
By the way, the reason why it doesn't build on Windows is because the CI system couldn't download the build environment, not because the code doesn't compile.

**Comment by Krzmbrzl** [comment]
> By the way, the reason why it doesn't build on Windows is because the CI system couldn't download the build environment, not because the code doesn't compile.

Yeah that's an issue of how the whole thing currently works. I restarted the CI though and hopefully now it runs through :)

**Comment by felix91gr** [comment]
Update on me as tester: I have downloaded and listened to it myself, but had some weird hardware problems? So couldn't really say much on it. ~Tomorrow is Friday, no wait, tomorrow is Thursday. This whole pandemic has made me lose the sense of time...~ Anywho... on Friday my friend should have more free time to help me test it.

Btw, I have a couple of questions just so we're on the same page... 

1. This PR makes it so activating RNNoise disables the normal noise cancellation feature, right? So now the noise cancellation is None XOR RNNoise XOR The "noise cancellation" slider.
2. What is the relationship of this and the Echo Cancellation feature, in the audio pipeline? 
3. Should I test all of this without echo cancellation? Should I test it with?
4. If I should test it _with_, then... should I pick Multichannel or Mixed? If I remember correctly, one of the two was sorta broken. I think it was the Multichannel one.

@fedetft

**Comment by fedetft** [comment]
1. the box "RNNoise" when ticked disables speex's noise canceller and makes the two sliders in the "Audio Processing" section useless. Fixes to the UI to make it clearer are not yet implemented.
2. **Before** this PR, enabling RNNoise did the following to the audio samples: first RNNoise would run, then the echo canceller would run on the output of RNNoise, and then speex's noise canceller would run on the output of the echo canceller. Such a nonstandard setup with an echo canceller in the middle of two noise cancellers makes no sense, and in addition RNNoise could impair the echo canceller performance.
This PR moves the noise canceller (either RNNoise or speex) always **after** the echo canceller.
3. The PR should be tested with the echo canceller to maximise coverage of the changes.
4. It shouldn't matter whether mixed or multichannel. I'm using mixed but that's because my audio setup doesn't benefit from multichannel (I use computer speakers/echo-prone headphones)

**Comment by felix91gr** [comment]
Okay, we see no changes. The change seems to be at least as good as the original on my side :)

**Comment by fedetft** [comment]
A lot of time has passed and I don't remember exactly what's missing to accepting this PR. @Krzmbrzl are you planning to bring the new flag up till the UI code (https://github.com/mumble-voip/mumble/pull/4212#issuecomment-635159785)?

**Comment by Krzmbrzl** [comment]
Sorry for the delayed response - I was under the impression that I have responded already...

The only thing really missing were testers that confirmed that RNNoise still works as expected. I was kinda hoping to get some more feedback but I guess as it is we can assume that it works just fine.
Thus I'll look into the UI stuff and add that to this PR and then I think we can merge it. Given that I am currently busy revamping my plugin PR though I think it could take me a week or so before I get to work on this PR :shrug:

**Comment by TredwellGit** [comment]
Just tested this. Disabling the Speex noise canceller when RNNoise is enabled makes it worse.

File: src/mumble/AudioInput.cpp
Comment (TredwellGit) [inline src/mumble/AudioInput.cpp:821]: Unnecessary comment?

Comment (fedetft) [reply src/mumble/AudioInput.cpp:821]: That comment will replace the line below once the UI code change is done and 	`bool bDenoise;` will be replaced with `NoiseCancel noiseCancel;` in `Struct settings`.

**Comment by fedetft** [comment]
@TredwellGit I know it may be hard to describe audio quality, but could you elaborate more about how it makes it worse?
I've been testing on my machine this branch all this time and the echo and noise canceller seem to work fine to me.

**Comment by TredwellGit** [comment]
RNNoise without Speex lets more noise, such as keyboard and mouse, through than RNNoise with Speex.

**Comment by fedetft** [comment]
So you basically would want to have the option to enable them both?
That can be done, after all the main reason for this patch is to order the DSP stuff in a reasonable way, and enabling both was not considered just because there seemed to be no use case for that, and it seemed just a waste of CPU cycles.

**Comment by fedetft** [comment]
@TredwellGit I added the option to enable them both to the DSP chain, but to test it out you'll need to wait for the UI change.

**Comment by TredwellGit** [comment]
Additionally, I would appreciate the ability to toggle all the Speex preprocessor parameters individually. A while back I compiled Mumble with all noise processing removed except Speex automatic gain control; in low noise environments it was very significantly better.

**Comment by Krzmbrzl** [comment]
> Additionally, I would appreciate the ability to toggle all the Speex preprocessor parameters individually.

I guess that's beyond the scope of this PR though :point_up: 
In general I think this could be a nice thing to have for experienced users...

**Comment by fedetft** [comment]
I agree with @Krzmbrzl, let's try to keep this PR focused to having a more structured noise cancel chain

**Comment by Krzmbrzl** [comment]
I (finally) added the UI component of this PR:
Preview image

Furthermore I adapted your commits in order to align with our new commit guidelines @fedetft 
That means that I had to rebase andforce-push your branch. Thus if you want to pull the changes, you'll have to do a `git pull -f`.

I think that this can then be merged pretty much as soon as the CI passes. In case someone has objections against that, I'll wait a day or two before actually merging this PR though :point_up:

File: src/mumble/AudioInput.cpp
Comment (davidebeatrici) [inline src/mumble/AudioInput.cpp:830]: This should be fixed in the commit that adds the message.

File: src/mumble/AudioInput.cpp
Comment (davidebeatrici) [inline src/mumble/AudioInput.cpp:825]:
```suggestion
		if (!denoiseState || iFrameSize != 480) {
```

**Review by davidebeatrici — APPROVED** [review]

**Review by Krzmbrzl — APPROVED** [review]

**Comment by Krzmbrzl** [comment]
@fedetft thank you very much for your contribution! :+1:

**Pull request merged**
